Account chooser dialog problem with some Android versions

Hi, first of all, thanks a lot for this library, it's used in PhoneTrack-Android. Luv'it.

So far, everything works fine with Android 4.1, 5.0, 8.0, 8.1 and 9.0.

I'm testing with freshly installed Nexus 4 virtual phones without google services but have confirmation it works fine on my Nexus 5x (LineageOS 15, Android 8.1).

I'm having problems with Android 6, 7.0 and 7.1 just after choosing an account (`AccountImporter.pickNewAccount()`)

With v0.3.0 It was apparently making Nextcloud Files app crash. Here are the corresponding logs: 
```
03-30 01:46:33.888  2104  2104 E AndroidRuntime: FATAL EXCEPTION: main
03-30 01:46:33.888  2104  2104 E AndroidRuntime: Process: com.nextcloud.client, PID: 2104
03-30 01:46:33.888  2104  2104 E AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{com.nextcloud.client/com.owncloud.android.ui.activity.SsoGrantPermissionActivity}: java.lang.NullPointerException: Attempt to read from field 'java.lang.String android.accounts.Account.name' on a null object reference
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2665)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2726)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.ActivityThread.-wrap12(ActivityThread.java)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1477)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.os.Handler.dispatchMessage(Handler.java:102)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.os.Looper.loop(Looper.java:154)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.ActivityThread.main(ActivityThread.java:6119)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at java.lang.reflect.Method.invoke(Native Method)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
03-30 01:46:33.888  2104  2104 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to read from field 'java.lang.String android.accounts.Account.name' on a null object reference
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at com.owncloud.android.ui.activity.SsoGrantPermissionActivity.onCreate(SsoGrantPermissionActivity.java:106)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.Activity.performCreate(Activity.java:6679)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1118)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2618)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    ... 9 more
03-30 01:46:33.891  1673  1803 W ActivityManager:   Force finishing activity com.nextcloud.client/com.owncloud.android.ui.activity.SsoGrantPermissionActivity
03-30 01:46:33.914  1673  1803 W ActivityManager:   Force finishing activity net.eneiluj.nextcloud.phonetrack.dev/net.eneiluj.nextcloud.phonetrack.android.activity.SettingsActivity
```

Hanging around in the issues, I discovered I could include `-SNAPSHOT` version instead of v0.3.0. Problem is still there in a less radical form :wink: : I get an alert dialog telling me `You have to accept the requested permissions to use the single sign-on feature.` Nothing in the logs. Is it possible I did something wrong? It seems to be related to `SsoGrantPermissionActivity`. Any idea?

Ow BTW, same thing is happening with News app, there's a crash. I guess it's using SSO lib v0.3.0. Same symptoms as PhoneTrack.

--- a/README.md
+++ b/README.md
@@ -67,6 +67,13 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
     }
 }
 
+@Override
+public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+    super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+
+    AccountImporter.onRequestPermissionsResult(requestCode, permissions, grantResults, this);
+}
+
 // Complete example: https://github.com/nextcloud/news-android/blob/master/News-Android-App/src/main/java/de/luhmer/owncloudnewsreader/LoginDialogFragment.java
 ```
 
@@ -250,7 +257,7 @@ if (VersionCheckHelper.verifyMinVersion(context, MIN_NEXTCLOUD_FILES_APP_VERSION
 }
 ``` 
 
-## Nextcloud Conference 2018 Talk
+## Nextcloud Conference 2018 Talk (5min)
 
 [![Nextcloud Single Sign On for Android David Luhmer](https://img.youtube.com/vi/gnLOwmrJLUw/0.jpg)](https://www.youtube.com/watch?v=gnLOwmrJLUw)
 

--- a/src/main/java/com/nextcloud/android/sso/AccountImporter.java
+++ b/src/main/java/com/nextcloud/android/sso/AccountImporter.java
@@ -19,6 +19,7 @@
 
 package com.nextcloud.android.sso;
 
+import android.Manifest;
 import android.accounts.Account;
 import android.accounts.AccountManager;
 import android.app.Activity;
@@ -28,10 +29,12 @@ import android.content.Context;
 import android.content.Intent;
 import android.content.SharedPreferences;
 import android.content.pm.PackageManager;
+import android.os.Build;
 import android.os.Bundle;
 import android.util.Log;
 import android.widget.Toast;
 
+import com.nextcloud.android.sso.exceptions.AndroidGetAccountsPermissionNotGranted;
 import com.nextcloud.android.sso.exceptions.NextcloudFilesAppAccountNotFoundException;
 import com.nextcloud.android.sso.exceptions.NextcloudFilesAppAccountPermissionNotGrantedException;
 import com.nextcloud.android.sso.exceptions.NextcloudFilesAppNotInstalledException;
@@ -44,7 +47,11 @@ import java.io.IOException;
 import java.util.ArrayList;
 import java.util.List;
 
+import androidx.core.app.ActivityCompat;
+import androidx.core.content.ContextCompat;
 import androidx.fragment.app.Fragment;
+
+import io.reactivex.annotations.NonNull;
 
 import static android.app.Activity.RESULT_CANCELED;
 import static android.app.Activity.RESULT_OK;
@@ -60,6 +67,7 @@ public class AccountImporter {
 
     public static final int CHOOSE_ACCOUNT_SSO = 4242;
     public static final int REQUEST_AUTH_TOKEN_SSO = 4243;
+    public static final int REQUEST_GET_ACCOUNTS_PERMISSION = 4244;
 
     private static SharedPreferences SHARED_PREFERENCES;
 
@@ -67,7 +75,9 @@ public class AccountImporter {
         return findAccounts(context).size() > 0;
     }
 
-    public static void pickNewAccount(Activity activity) throws NextcloudFilesAppNotInstalledException {
+    public static void pickNewAccount(Activity activity) throws NextcloudFilesAppNotInstalledException, AndroidGetAccountsPermissionNotGranted {
+        checkAndroidAccountPermissions(activity);
+
         if (appInstalledOrNot(activity, "com.nextcloud.client")) {
             Intent intent = AccountManager.newChooseAccountIntent(null, null, new String[]{"nextcloud"},
                     true, null, null, null, null);
@@ -77,13 +87,34 @@ public class AccountImporter {
         }
     }
 
-    public static void pickNewAccount(Fragment fragment) throws NextcloudFilesAppNotInstalledException {
+    public static void pickNewAccount(Fragment fragment) throws NextcloudFilesAppNotInstalledException, AndroidGetAccountsPermissionNotGranted {
+        checkAndroidAccountPermissions(fragment.getContext());
+
         if (appInstalledOrNot(fragment.getContext(), "com.nextcloud.client")) {
             Intent intent = AccountManager.newChooseAccountIntent(null, null, new String[]{"nextcloud"},
                     true, null, null, null, null);
             fragment.startActivityForResult(intent, CHOOSE_ACCOUNT_SSO);
         } else {
             throw new NextcloudFilesAppNotInstalledException();
+        }
+    }
+
+    public static void requestAndroidAccountPermissionsAndPickAccount(Activity activity) {
+        ActivityCompat.requestPermissions(activity, new String[]{Manifest.permission.GET_ACCOUNTS}, REQUEST_GET_ACCOUNTS_PERMISSION);
+    }
+
+    private static void checkAndroidAccountPermissions(Context context) throws AndroidGetAccountsPermissionNotGranted {
+        // https://developer.android.com/reference/android/accounts/AccountManager#getAccountsByType(java.lang.String)
+        // Caller targeting API level below Build.VERSION_CODES.O that have not been granted the Manifest.permission.GET_ACCOUNTS permission,
+        // will only see those accounts managed by AbstractAccountAuthenticators whose signature matches the client.
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+            // Do something for lollipop and above versions
+            if (ContextCompat.checkSelfPermission(context, Manifest.permission.GET_ACCOUNTS) != PackageManager.PERMISSION_GRANTED) {
+                Log.e(TAG, "Permission not granted yet!");
+                throw new AndroidGetAccountsPermissionNotGranted();
+            } else {
+                Log.d(TAG, "Permission granted!");
+            }
         }
     }
 
@@ -101,7 +132,7 @@ public class AccountImporter {
     // Find all currently installed nextcloud accounts on the phone
     public static List<Account> findAccounts(final Context context) {
         final AccountManager accMgr = AccountManager.get(context);
-        final Account[] accounts = accMgr.getAccounts();
+        final Account[] accounts = accMgr.getAccountsByType("nextcloud");
 
         List<Account> accountsAvailable = new ArrayList<>();
         for (final Account account : accounts) {
@@ -201,6 +232,17 @@ public class AccountImporter {
                     SingleSignOnAccount singleSignOnAccount = extractSingleSignOnAccountFromResponse(data, context);
                     callback.accountAccessGranted(singleSignOnAccount);
                     break;
+                case REQUEST_GET_ACCOUNTS_PERMISSION:
+                    try {
+                        if(activity != null) {
+                            pickNewAccount(activity);
+                        } else {
+                            pickNewAccount(fragment);
+                        }
+                    } catch (NextcloudFilesAppNotInstalledException  | AndroidGetAccountsPermissionNotGranted e) {
+                        UiExceptionManager.showDialogForException(context, e);
+                    }
+                    break;
                 default:
                     break;
             }
@@ -220,10 +262,48 @@ public class AccountImporter {
                         Log.e(TAG, e.getMessage());
                     }
                     break;
+                case REQUEST_GET_ACCOUNTS_PERMISSION:
+                    UiExceptionManager.showDialogForException(context, new AndroidGetAccountsPermissionNotGranted());
+                    break;
                 default:
                     break;
             }
         }
+    }
+
+    public static void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults, Activity activity) {
+        onRequestPermissionsResult(requestCode, permissions, grantResults, activity, null);
+    }
+
+    public static void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults, Fragment fragment) {
+        onRequestPermissionsResult(requestCode, permissions, grantResults, null, fragment);
+    }
+
+    private static void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults, Activity activity, Fragment fragment) {
+        Context context = (activity != null) ? activity : fragment.getContext();
+
+        switch (requestCode) {
+            case REQUEST_GET_ACCOUNTS_PERMISSION:
+                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                    // Permissions have been granted.. start pick account dialog
+                    try {
+                        if (activity != null) {
+                            pickNewAccount(activity);
+                        } else {
+                            pickNewAccount(fragment);
+                        }
+                    } catch (NextcloudFilesAppNotInstalledException | AndroidGetAccountsPermissionNotGranted e) {
+                        UiExceptionManager.showDialogForException(context, e);
+                    }
+                } else {
+                    // user declined the permission request..
+                    UiExceptionManager.showDialogForException(context, new AndroidGetAccountsPermissionNotGranted());
+                }
+                break;
+            default:
+                break;
+        }
+
     }
 
     public static void handleFailedAuthRequest(Intent data) throws SSOException {

--- a/src/main/java/com/nextcloud/android/sso/exceptions/AndroidGetAccountsPermissionNotGranted.java
+++ b/src/main/java/com/nextcloud/android/sso/exceptions/AndroidGetAccountsPermissionNotGranted.java
@@ -1,0 +1,17 @@
+package com.nextcloud.android.sso.exceptions;
+
+import android.content.Context;
+
+import com.nextcloud.android.sso.R;
+import com.nextcloud.android.sso.model.ExceptionMessage;
+
+public class AndroidGetAccountsPermissionNotGranted extends SSOException {
+
+    @Override
+    public void loadExceptionMessage(Context context) {
+        this.em = new ExceptionMessage(
+             context.getString(R.string.android_get_accounts_permission_not_granted_exception_title),
+             context.getString(R.string.android_get_accounts_permission_not_granted_exception_message)
+        );
+    }
+}

--- a/src/main/java/com/nextcloud/android/sso/exceptions/CurrentAccountNotFoundException.java
+++ b/src/main/java/com/nextcloud/android/sso/exceptions/CurrentAccountNotFoundException.java
@@ -29,8 +29,8 @@ public class CurrentAccountNotFoundException extends SSOException {
     @Override
     public void loadExceptionMessage(Context context) {
         this.em = new ExceptionMessage(
-            em.title   = context.getString(R.string.current_account_not_found_exception_title),
-            em.message = context.getString(R.string.current_account_not_found_exception_message)
+            context.getString(R.string.current_account_not_found_exception_title),
+            context.getString(R.string.current_account_not_found_exception_message)
         );
     }
 }

--- a/src/main/res/values/strings.xml
+++ b/src/main/res/values/strings.xml
@@ -3,6 +3,9 @@
     <string name="no_current_account_selected_exception_title">Warning</string>
     <string name="no_current_account_selected_exception_message">No account selected yet</string>
 
+    <string name="android_get_accounts_permission_not_granted_exception_title">Error</string>
+    <string name="android_get_accounts_permission_not_granted_exception_message">In order to use the Nextcloud Single-Sign-On on this device, you need to grant permission to access your accounts</string>
+
     <string name="current_account_not_found_exception_title">Warning</string>
     <string name="current_account_not_found_exception_message">Selected account was not found. Something went wrong..</string>
 

